content/starter-configs/examples/ruby-on-rails.md

```diff
@@ -0,0 +1,108 @@
+    #   3. Routes requests to the Preview URL to this service
+    default: true
+
+    # Don't build the Ruby service until the database service is started
+    depends: database
+
+    # A set of commands to run while building this service
+    commands:
+      # Commands that set up the basic preview infrastructure
+      init:
+        # Check for updates to apt-get, and then use it to install Ruby
+        - apt-get update
+        - apt-get install ruby ruby-dev
+
+        # Install a minimal set of PostgreSQL binaries and headers to interact with the database
+        - apt-get install libpq-dev
+
+        # Install the gems required for the project
+        - gem update --system
+        - gem install bundler
+        - bundle install
+
+        # Clean up apt artifacts to keep the Preview small
+        - apt-get clean
+        - rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+
+      # Commands that import files, databases,  or other assets. When an
+      # existing preview is refreshed, the build workflow starts here,
+      # skipping the init step, because the results of that step will
+      # already be present.
+      update:
+        # A common practice in many Rails projects is to run this command to
+        # set up a database. This step is why the 'database' service has an 'init'
+        # command to give the default Tugboat user permission to create the database.
+        # If you're not creating a database here, you don't need the 'init' and
+        # 'online' commands in the `database` service. If you're connecting to a remote
+        # database, you won't need this.
+        - bin/rails db:setup
+
+      # Commands that should happen every time the Preview starts.
+      start:
+        # Start the Rails server
+        - bin/rails s -b 0.0.0.0 -p 80
+
+  # What to call the service hosting the database. This name also acts as the
+  # hostname to access the service by from the 'rails' service.
+  database:
+    # Use the latest available version of postgres by not specifying a
+    # version
+    image: tugboatqa/postgres
+
+    # A set of commands to run while building this service
+    commands:
+      # Commands that import files, databases, or other assets. When an
+      # existing preview is refreshed, the build workflow starts here,
+      # skipping the init step, because the results of that step will
+      # already be present.
+      init:
+        # Give the default Tugboat user the permission to create a database.
+        # This is only required if using 'rails db:setup'
+        - psql -U postgres -d tugboat -c "ALTER USER tugboat WITH CREATEDB;"
+
+      # Tugboat runs these commands once, after the Preview comes online for the first time
+      online:
+        # For security purposes, remove the extra permissions we gave the
+        # Tugboat user to create the database in 'init'
+        - psql -U postgres -d tugboat -c "ALTER USER tugboat WITH NOCREATEDB;"
```

Comment:
I've been thinking about it, and I don't think this is entirely necessary. We could probably even give the tugboat user `CREATEDB` permissions by default in our image. Postgres permissions are strange.

Reply:
Hmm. I don't know enough about the security implications either way, or what users would "expect," permissions-wise, in a Postgres image. When we chatted on Friday, I thought that was what Kay was going to ask about, but he seemed satisfied with your explanation of why and said that seems reasonable. I'd be curious how other people approach this.

Reply:
It certainly doesn't hurt to drop the privileges. But, the change won't stick if someone clones or resets the preview, so it feels a little misleading to recommend it "for security purposes"

I do think it's OK for the tugboat user to retain CREATEDB permissions for the life of a preview, which would at least make the config.yml a little easier to grok.

If this were a production instance I would have different opinions.

Reply:
Alrighty, I've removed the `online` command related to dropping `CREATEDB` permissions. We can drop the `init` step entirely if we add the `CREATEDB` permissions to the Postgres image, but I don't necessarily want to delay this until that happens.

Reply:
Agreed. Let's leave that bit in. It should work regardless of whether the tugboat ends up with CREATEDB permissions in the image.